tools/c7n_azure/c7n_azure/resources/storage.py

```diff
@@ -104,3 +115,131 @@ def _query_rules(self, resource):
         resource_rules = set([IPNetwork(r['value']) for r in ip_rules])
 
         return resource_rules
+
+
+@Storage.filter_registry.register('storage-diagnostic-settings')
+class StorageDiagnosticSettingsFilter(ValueFilter):
+    BLOB_TYPE = 'blob'
+    QUEUE_TYPE = 'queue'
+    TABLE_TYPE = 'table'
+    FILE_TYPE = 'file'
+
+    def __init__(self, data, manager=None):
+        super(StorageDiagnosticSettingsFilter, self).__init__(data, manager)
+        self.storage_type = data['storage_type']
+
+    schema = type_schema('storage-diagnostic-settings',
+                         rinherit=ValueFilter.schema,
+                         storage_type={
+                             'type': 'string',
+                             'enum': [BLOB_TYPE, QUEUE_TYPE, TABLE_TYPE, FILE_TYPE]},
+                         required=['storage_type'],
+                         )
+
+    def process(self, resources, event=None):
+        futures = []
+        results = []
+        session = local_session(self.manager.session_factory)
+        # Process each resource in a separate thread, returning all that pass filter
+        with self.executor_factory(max_workers=3) as w:
+            for resource_set in chunks(resources, 20):
+                futures.append(w.submit(self.process_resource_set, resource_set, session))
+
+            for f in as_completed(futures):
+                if f.exception():
+                    self.log.warning(
+                        "Storage diagnostic settings filter error: %s" % f.exception())
+                    continue
+                else:
+                    results.extend(f.result())
+
+            return results
+
+    def process_resource_set(self, resources, session):
+        matched = []
+        for resource in resources:
+            settings = json.loads(jsonpickle.encode(self.get_settings(resource, session)))
```

Comment:
Can we cache settings?

Reply:
I'm going to update how this works so we won't need to cache.